python/cudf_polars/cudf_polars/experimental/explain.py

```diff
@@ -0,0 +1,230 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Explain logical and physical plans."""
+
+from __future__ import annotations
+
+import functools
+from collections.abc import MutableMapping
+from itertools import groupby
+from typing import TYPE_CHECKING
+
+from cudf_polars.dsl.ir import (
+    IR,
+    GroupBy,
+    Join,
+    Projection,
+    Scan,
+    Select,
+    Sort,
+    Union,
+)
+from cudf_polars.dsl.translate import Translator
+from cudf_polars.experimental.base import PartitionInfo
+from cudf_polars.experimental.io import SplitScan
+from cudf_polars.experimental.parallel import lower_ir_graph
+from cudf_polars.utils.config import ConfigOptions
+
+if TYPE_CHECKING:
+    import polars as pl
+
+
+def explain_query(
+    q: pl.LazyFrame, engine: pl.GPUEngine, *, physical: bool = True
+) -> str:
+    """
+    Return a formatted string representation of the IR plan.
+
+    Parameters
+    ----------
+    q : pl.LazyFrame
+        The LazyFrame to explain.
+    engine : pl.GPUEngine
+        The configured GPU engine to use.
+    physical : bool, default True
+        If True, show the physical (lowered) plan.
+        If False, show the logical (pre-lowering) plan.
+
+    Returns
+    -------
+    str
+        A string representation of the IR plan.
+    """
+    config = ConfigOptions.from_polars_engine(engine)
+    ir = Translator(q._ldf.visit(), engine).translate_ir()
+
+    if physical:
+        lowered_ir, partition_info = lower_ir_graph(ir, config)
+        return _format_ir_tree(lowered_ir, partition_info)
+    else:
+        return _format_ir_tree(ir)
+
+
+def _format_ir_tree(
+    ir: IR,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    header = _format_ir(ir, partition_info, offset=offset)
+
+    children_strs = [
+        _format_ir_tree(child, partition_info, offset=offset + "  ")
+        for child in ir.children
+    ]
+
+    formatted = []
+    for line, group in groupby(children_strs):
+        count = sum(1 for _ in group)
+        formatted.append(line)
+        if count > 1:
+            formatted.append(f"{offset}  (repeated {count} times)\n")
+
+    return header + "".join(formatted)
+
+
+@functools.singledispatch
+def _format_ir(
+    ir: IR,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    count = partition_info[ir].count if partition_info else None
+    header = f"{offset}{type(ir).__name__.upper()}"
```

Comment:
Maybe the default repr should include the schema, and then we don't need to define anything for `Projection` and `Select`?  It may also make sense to avoid printing the full schema for "wide" tables (e.g. >6 columns or something).

Reply:
Good ideas, done. Added tests for wide tables in https://github.com/rapidsai/cudf/pull/18708/commits/ff2a670eec361dec888139b5968228b1545bde7e